File: datastruct/trie/Trie.java
import java.util.TreeMap;

public class Trie {
    private class Node{
        boolean isWord;
        TreeMap<Character, Node> next;

        public Node(){
            this(false);
        }

        public Node(boolean isWord){
            this.isWord = isWord;
            next = new TreeMap<> ();
        }
    }

    private Node root;
    private int size;

    public Trie() {
        root = new Node();
        size = 0;
    }

    // 获得 Trie 中存储的单词数量
    public int getSize() {
        return size;
    }

    // 向 Trie 中添加新的单词
    public void add(String word){
        Node cur = root;
        for(int i = 0; i < word.length();i++){
            char c = word.charAt(i);
            if (cur.next.get(c) == null)
                cur.next.put(c, new Node());
            cur = cur.next.get(c);
        }

        if(!cur.isWord){
            cur.isWord = true;
            size++;
        }

    }

    // 查询 Trie 中是否包含单词 word
    public boolean contains(String word){
        Node cur = root;
        for (int i = 0; i < word.length(); i++) {
            char c = word.charAt(i);
            if (cur.next.get(c) == null) {
                return false;
            }
            cur = cur.next.get(c);
        }

        return cur.isWord;
    }

    // 查询 Trie 中是否包括以 prefix 为前缀的单词
    public boolean isPrefix(String prefix){
        Node cur = root;
        for(int i = 0; i < prefix.length(); i++){
            char c = prefix.charAt(i);
            if(cur.next.get(c) == null)
                return false;
            cur = cur.next.get(c);
        }
        return true;
    }

}

File: datastruct/trie/Trie.py
class Node(object):

        def __init__(self, is_word=False):
            self.next = {}
            self.is_word = is_word

class Trie(object):

    def __init__(self):
        self.__root = Node()
        self.__size = 0

    def get_size(self):
        return self.__size

    def add(self, word:str):
        '''
            向 Trie 中添加单词 word
        '''
        cur = self.__root
        for c in word:
            if(cur.next.get(c) is None):
                cur.next.update({c:Node()})
            cur = cur.next.get(c)
        if cur.is_word == False:
            cur.is_word = True
            self.__size += 1

    def contain(self, word:str):
        '''
            查看 Trie 中是否包含单词 word
        '''
        cur = self.__root
        for c in word:
            if(cur.next.get(c) is None):
                return False
            cur = cur.next.get(c)
        return cur.is_word
    
    def is_prefix(self, prefix:str):
        '''
            查看 Trie 中是否有以 prefix 为前缀的单词
        '''
        cur = self.__root
        for c in prefix:
            if(cur.next.get(c) is None):
                return False
            cur = cur.next.get(c)
        
        return True

    